processing/src/main/java/org/apache/druid/query/operator/SegmentToRowsAndColumnsOperator.java

```diff
@@ -44,7 +46,13 @@ public Closeable goOrContinue(Closeable continuation, Receiver receiver)
   {
     try (final CloseableShapeshifter shifty = segment.as(CloseableShapeshifter.class)) {
       if (shifty == null) {
-        throw new ISE("Segment[%s] cannot shapeshift", segment.getClass());
+        final Class<? extends Segment> classOfSegment;
+        if (segment instanceof ReferenceCountingSegment) {
```

Comment:
this seems like an abstraction leak - I think this should be the responsibility of `Segment` somethow.

Wouldn't it be better to implement the `toString()` methods so that it contains some details - and class name of the base segment if its a `ReferenceCountingSegment` ?